R/checkPythonDeps.R

```diff
@@ -0,0 +1,350 @@
+#' @param strict Strict comparison of version components (e.g. 3.10 != 3.10.0)
+#' @return TRUE operator is satisfied, FALSE otherwise
+comparePythonVersions <- function(operator, first, second, strict = TRUE) {
+  # Compare individual components of the version strings in decreasing order of importance
+  for (component in c("major", "minor", "patch", "releaseVersion")) {
+    # If strict comparison is not required and the component is NA in either version, skip the comparison
+    if (!strict && (is.na(first[[component]]) || is.na(second[[component]]))) next
+    # Choosing -1 for NA means missing version components are considered less than 0.
+    # Hence, a version like 3.10 (= 3.10.NA) would be considered less than 3.10.0.
+    v1 <- ifelse(is.na(first[[component]]), -1, first[[component]])
+    v2 <- ifelse(is.na(second[[component]]), -1, second[[component]])
+    # Since we're comparing in decreasing order of importance, we can continue if components are equal ..
+    if (v1 == v2) next
+    # .. or immediately return the result of the comparison if they are not equal
+    return(switch(operator,
+      ">=" = v1 >= v2,
+      "==" = v1 == v2,
+      "<=" = v1 <= v2,
+      "!=" = v1 != v2,
+      ">" = v1 > v2,
+      "<" = v1 < v2,
+      stop("Invalid operator '", operator, "'. Must be either '==', '!=', '>', '>=', '<' or '<='.")
+    ))
+  }
+  # If all components are equal or NA (in non-strict mode), return the result of the equality comparison
+  return(switch(operator,
+    ">=" = TRUE,
+    "==" = TRUE,
+    "<=" = TRUE,
+    FALSE
+  ))
+}
+
+#' Create a named list representation of a Python dependency
+#'
+#' This function creates a named list representing a dependency on a particular Python package
+#'
+#' @param name Name of the dependency
+#' @param operator Comparison operator. One of '==', '!=', '>', '>=', '<' or '<='
+#' @param version Named list representation of the Python version of the dependency
+#' @param build Build version of the dependency
+#' @param repo Repository URL of the dependency
+createPythonDependency <- function(name = "", operator = "", version = createPythonVersion(), build = "", repo = "") {
+  return(list(
+    name = name,
+    operator = operator,
+    version = version,
+    build = build,
+    repo = repo
+  ))
+}
+
+#' Extract Python dependency from dependency string
+#'
+#' This function extracts the package name, relationship operator and version from a typical Python dependency string
+#' (e.g. "numpy==1.26.4") and returns a named list representing the dependency. Dependencies on specific repositories
+#' are also supported (e.g. "climate-assessment @ <URL to git repo>")
+#'
+#' @param depString Dependency string such as "numpy<=2.0"
+#' @param style Style of the dependency string. Either "pip" or "conda"
+#' @return A named list representation of the Pzthon package name and version
+extractPythonDependency <- function(depString, style = "pip") {
```

Comment:
This again feels like it should already exist in a python module. Writing this kind of code from scratch seems very risky to me.

Reply:
You're right, Python can do this, however we need R to do it, too 😬 This function is called before we can rely on using Python in REMIND